### PR TITLE
[WIP] Swift 3.0

### PR DESCRIPTION
**Pull request by marcelofabri**
This adds support to Swift 3.0. It's still WIP:
- [x] Update core
- [x] Update core tests
- [x] Update RAC
- [x] Update RAC tests
- [ ] Update podspec
- [x] Point OHHTTPStubs to https://github.com/AliSoftware/OHHTTPStubs/pull/201 or wait until that is merged
- [x] ~~Update tests to run on iOS 10: they seem to be failing on the Simulator with the same issue mentioned here: https://forums.developer.apple.com/thread/51071 and http://stackoverflow.com/questions/38456471/secitemadd-always-returns-error-34018-in-xcode-8-beta-4-in-ios-10-simulator~~ I think this should be done on a later PR, since it'll probably need to create an app target so the unit tests could be run on a host app (with an Entitlements file)
- [ ] Lock dependencies to releases (some of them are referring to branches)

### CODE REVIEW

File: Heimdallr/Core/URLRequestExtensions.swift [hunk 1]
```diff
@@ -109,45 +107,13 @@ public extension NSMutableURLRequest {
...
-
-            while index != string.endIndex {
-                let startIndex = index
-                let endIndex = index.advancedBy(batchSize, limit: string.endIndex)
-                let range = startIndex..<endIndex
-
-                let substring = string.substringWithRange(range)
-
-                escaped += substring.stringByAddingPercentEncodingWithAllowedCharacters(allowedCharacterSet) ?? substring
+        var allowedCharacterSet = CharacterSet.urlQueryAllowed
```
Comment (marcelofabri) [inline Heimdallr/Core/URLRequestExtensions.swift]: I've updated this from Alamofire code. It seems that they are dropping iOS 8.x on the new release. Maybe we should do the same? (If not, this needs to be reverted and converted to Swift 3.0)

**Comment by marcelofabri** [comment]
Also, I chose to not mark the classes as `open` as the converter suggested, but that may be discussible

**Comment by marcelofabri** [comment]
I think Circle hasn't update to the GM yet

File: Cartfile.resolved [hunk 2]
```diff
@@ -1,6 +1,6 @@
...
-github "AliSoftware/OHHTTPStubs" "f86d9825f1b4fdb91620fd18e2a017fc5ce0e7ba"
-github "Quick/Quick" "81d2a7bd95ef91e2604ee0431bba6fe59ad662dc"
-github "antitypical/Result" "3.0.0-alpha.4"
+github "Quick/Nimble" "db706fc1d7130f6ac96c56aaf0e635fa3217fe57"
+github "AliSoftware/OHHTTPStubs" "5.2.1-swift3"
+github "Quick/Quick" "1503fb019d72417d5d6e4fbebdbaa03c9e4a125f"
+github "antitypical/Result" "3.0.0"
 github "jspahrsummers/xcconfigs" "0.9"
-github "ReactiveCocoa/ReactiveCocoa" "v4.2.2"
+github "ReactiveCocoa/ReactiveCocoa" "v4.0.0-alpha.3"
```
Comment (marcelofabri) [inline Cartfile.resolved]: for some reason, this was changed when I updated Nimble

¯\_(ツ)_/¯

**Comment by marcelofabri** [comment]
So, the Mac and iOS builds are passing, but the Reactive Cocoa isn't. From what I've seen, 4.x won't support Swift 3, so we'd need to wait for 5.0. I have no idea if `master` is currently stable or not as I have little familiarity with RC.

What do you guys think?

File: Heimdallr/Core/URLRequestExtensions.swift (hunk 1 above)
Comment (hffmnn) [inline Heimdallr/Core/URLRequestExtensions.swift]: From my point of view it is ok to drop iOS 8 support. If someone needs it, we are open for a Pull Request :)

**Comment by SusannProszak** [comment]
hello @marcelofabri 
thanks for your pr. 👍 
we prefer to wait until ReactiveCocoa releases swift 3.0 support. Until then I would leave the pr in [WIP].

**Comment by judi0713** [comment]
So, is there any plan to migrate Heimdallr to Swift 3.0?
I already use Xcode 8😔Recently I want to update my project to support swift 3.0.

**Comment by marcelofabri** [comment]
Should we use https://github.com/ReactiveCocoa/ReactiveSwift instead of RAC 4? If so, there's already a 1.0.0-alpha.1 release we can use.

**Comment by marcelofabri** [comment]
I've updated the dependencies now that Nimble and Quick have stable versions and also updated the code to use ReactiveSwift instead of ReactiveCocoa. 

However, I had to comment all the `Objective-C compatibility` section from `Heimdallr+ReactiveCocoa` because I didn't know how to migrate that.

**Comment by BabyAzerty** [comment]
Excuse me, I don't have much details into the project but as a user of Heimdallr, I would like to understand why is there - historically - blocking dependencies still here?
There was already a problem with the JSON parser before (Argo).
And now RAC.

Or is it that we can DL this branch, exclude RAC, and get things working?

**Comment by marcelofabri** [comment]
@BabyAzerty If you use CocoaPods, you can use my fork without the RAC subspec (it's optional), so you can migrate your project to Swift 3.0 while waiting this PR to be merged. Just change your Podfile:

```ruby
pod 'Heimdallr', git: 'https://github.com/marcelofabri/Heimdallr.swift.git', branch: 'swift-3.0'
```

While I can't speak for the maintainers, I think it'd be useful to make the RAC extension another repository to make it easier to manage, but I don't know how complex that would make things.

**Comment by bckr** [comment]
> I've updated the dependencies now that Nimble and Quick have stable versions and also updated the code to use ReactiveSwift instead of ReactiveCocoa.

Thank you so much for this and sorry for our late response here. We've been a bit busy the last couple of weeks but I would like to rekindle the discussion here again.

> However, I had to comment all the Objective-C compatibility section from Heimdallr+ReactiveCocoa because I didn't know how to migrate that.

I think if we want to have full compatibility with Objc and Swift we'll require both `ReactiveCocoa` and the `ReactiveObjCBridge` now because the repository [will be split up](https://github.com/ReactiveCocoa/ReactiveCocoa/pull/3118/files) into multiple smaller ones for the upcoming release that brings Swift 3 compatibility.

> While I can't speak for the maintainers, I think it'd be useful to make the RAC extension another repository to make it easier to manage, but I don't know how complex that would make things.

I like that idea! I believe it would make the project way easier to maintain. Since we are also internally moving away from heavy RAC usage in our project, it might be the right direction for `Heimdallr` as well. The reactive extensions can still live in a separate repository. I'll discuss it internally and comment here if we have made a final decision. Otherwise the Swift 3 migration would be blocked until we get a stable RAC 5.0 release.

**Comment by bckr** [comment]
We decided that we are fine with splitting up the repository and removing RAC as a dependency for this repository.

@marcelofabri Do you want to tackle that? I think you already provide a working Swift 3 fork without the RAC subspec? We'll create a second repository for the reactive extensions when we have a stable RAC release. Let me know if I can help you out in any way 😊

**Comment by marcelofabri** [comment]
@bckr I'm currently traveling, so it's unlikely that I'm able to work on this for the next ~2 weeks. The fork currently have a RAC subspec, but it's just not working 100%.

I've allowed edits from maintainers, so feel free to start from my fork if you wanna handle this.

**Comment by bckr** [comment]
@marcelofabri Will do! Thanks!

**Comment by Ingibjorg** [comment]
Hi all,
what is the ETD on this update? Heimdallr is the only library that prevents me from being able to update my project from Swift 2.3 to Swift 3.0. Xcode 8.2 will be the last release that will support Swift 2.3 so we should migrate quickly to Swift 3.0.

**Comment by jPaolantonio** [comment]
Is there anything I can do to help the Swift 3 migration?

**Pull request merged**